src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore
+{
+    internal sealed class NuGetRestoreCycleDetector
+    {
+        private readonly object _lock;
+        private readonly Queue<byte[]> _values;
+        private int counter;
+
+        /// <summary>
+        ///     Fixed size of the numbers of values to store.
+        /// </summary>
+        /// <remarks>
+        ///     This represents how deep to search for hash1 cycle.
+        /// </remarks>
+        public int Size { get; private set; }
+
+        public NuGetRestoreCycleDetector(int size)
+        {
+            _lock = new object();
+            _values = new Queue<byte[]>();
+            Size = size;
+        }
+
+        /// <summary>
+        ///     Validate if hash1 cycle exist. 
+        ///     If no cycle exist, then the hashValue is stored.
+        /// </summary>
+        /// <param name="hashValue">This is hashValue that is used to verify if it exists in previous restores</param>
+        /// <returns>True if it contains hash1 cycle, otherwise false</returns>
+        public bool ComputeCycleDetection(byte[] hashValue)
+        {
+            lock (_lock)
+            {
+                if (QueueContainsValue(hashValue))
+                {
+                    counter++;
+                    
+                    // Verify that a hashValue has repeted in almost all cases
+                    if (counter > Size)
+                    {
+                        counter = 0;
+                        ClearQueue();
+                        return true;
+                    }
+                }
+                else
+                {
+                    Add(hashValue);
```

Comment:
If the queue does not contain this value, I think the counter should be reset to zero.

Also the length of the queue should probably be capped. Otherwise the probability of finding the hash in the queue increases over time.

Reply:
done.